From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xVorttex <sigamatute@gmail.com>
Date: Tue, 25 Jul 2023 02:38:41 +0300
Subject: [PATCH] Helpers


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 74a9e27b8d3147876483743e1107632e332b0571..eff4090292d4c2e5a7fefd7555a39ecc8dadab15 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -52,6 +52,8 @@ import org.jetbrains.annotations.Nullable;
 public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
 
     // Paper start
+    void sendPacket(@NotNull Object packet);
+
     @Override
     default net.kyori.adventure.identity.@NotNull Identity identity() {
         return net.kyori.adventure.identity.Identity.identity(this.getUniqueId());
