From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xVorttex <sigamatute@gmail.com>
Date: Fri, 24 Nov 2023 20:25:33 +0200
Subject: [PATCH] Location libs


diff --git a/src/main/java/org/bukkit/Location.java b/src/main/java/org/bukkit/Location.java
index ed3be01cd9bfa1785e3f3be8bd2d70a293e86512..345903aceb2b08507245faf10f534a30e4694c90 100644
--- a/src/main/java/org/bukkit/Location.java
+++ b/src/main/java/org/bukkit/Location.java
@@ -130,6 +130,8 @@ public class Location implements Cloneable, ConfigurationSerializable, io.paperm
     @BsonIgnore
     @UndefinedNullability // Paper
     public World getWorld() {
+        if (this.world == null && this.worldName == null)
+            return null;
         World world = findWorld();
         Preconditions.checkArgument(world != null, "World unloaded");
         return world;
