From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xVorttex <sigamatute@gmail.com>
Date: Sat, 25 Nov 2023 14:29:26 +0200
Subject: [PATCH] Location bson


diff --git a/src/main/java/org/bukkit/Location.java b/src/main/java/org/bukkit/Location.java
index 345903aceb2b08507245faf10f534a30e4694c90..f0cec595a4ae01aece52927dbcc480dceb515365 100644
--- a/src/main/java/org/bukkit/Location.java
+++ b/src/main/java/org/bukkit/Location.java
@@ -137,6 +137,14 @@ public class Location implements Cloneable, ConfigurationSerializable, io.paperm
         return world;
     }
 
+    public String getWorldName() {
+        return worldName;
+    }
+
+    public void setWorldName(String worldName) {
+        this.worldName = worldName;
+    }
+
     @BsonIgnore
     private World findWorld() {
         World world = null;
