From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xVorttex <sigamatute@gmail.com>
Date: Sat, 25 Nov 2023 15:16:15 +0200
Subject: [PATCH] Helpers


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 8d963acdeb99cfc0b424b9b3bdbda40b1c73e7a2..0615a6777f062e5b2b2e84666ca6fddfdc64d746 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -52,6 +52,7 @@ import org.jetbrains.annotations.Nullable;
  * Represents a player, connected or not
  */
 public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
+    void sendPacket(@NotNull Object packet);
 
     // Paper start
     @Override
