From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xVorttex <sigamatute@gmail.com>
Date: Mon, 23 Sep 2024 20:49:33 +0300
Subject: [PATCH] Bson annotations


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index aac21ff5418f76b0d06b6a49a4021bde194cf1da..e018717724e934e174a0cfc15de24681c5745138 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -217,6 +217,12 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.firstPlayed = System.currentTimeMillis();
     }
 
+    @Override
+    public void sendPacket(@NotNull Object obj) {
+        if (obj instanceof Packet<?> packet)
+            getHandle().connection.send(packet);
+    }
+    
     public GameProfile getProfile() {
         return this.getHandle().getGameProfile();
     }
