From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xVorttex <sigamatute@gmail.com>
Date: Tue, 25 Jul 2023 02:39:52 +0300
Subject: [PATCH] Fixed unknown method


diff --git a/src/main/java/net/minecraft/network/protocol/game/ClientboundTeleportEntityPacket.java b/src/main/java/net/minecraft/network/protocol/game/ClientboundTeleportEntityPacket.java
index b1dbdc7984817134b196858b30ede45a7a85074a..fe2f6ac96ea8bc914108578ff67add756d22432b 100644
--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundTeleportEntityPacket.java
+++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundTeleportEntityPacket.java
@@ -23,7 +23,7 @@ public class ClientboundTeleportEntityPacket implements Packet<ClientGamePacketL
         this.z = vec3.z;
         this.yRot = (byte)((int)(entity.getYRot() * 256.0F / 360.0F));
         this.xRot = (byte)((int)(entity.getXRot() * 256.0F / 360.0F));
-        this.onGround = entity.isOnGround();
+        this.onGround = entity.onGround();
     }
 
     public ClientboundTeleportEntityPacket(FriendlyByteBuf buf) {
